File: reader.rb
require 'csv'
require 'benchmark'
require 'ruby-progressbar'

FILENAME = ARGV.size > 0 ? ARGV[0] : 'data.csv'

def memory_usage
  memory_before = `ps -o rss= -p #{Process.pid}`.to_i
  yield
  memory_after = `ps -o rss= -p #{Process.pid}`.to_i

  puts "Memory: #{((memory_after - memory_before) / 1024.0).round(2)} MB"
end

def time_spent
  time = Benchmark.realtime do
    yield
  end

  puts "Time: #{time.round(2)}"
end

memory_usage do
  time_spent do
    # count = File.foreach(FILENAME).inject(0) { |c, _| c + 1 }
    # bar = ProgressBar.create(throttle_rate: 0.2, total: count)

    CSV.foreach(FILENAME) do |row|
      # bar.incremen
    end
  end
end

File: reader.py
import os
import sys
import csv
import time
import functools
import progressbar
import psutil

csv_path = sys.argv[1] if len(sys.argv) > 1 else 'data.csv' 

examples = []

def example(fn):
    @functools.wraps(fn)
    def wrapped():
        try:
            print('Running: {}'.format(fn.__name__))

            start = time.time()

            fn()

            end = time.time()
            print('{:.2f}'.format(end - start))

            process = psutil.Process(os.getpid())
            print('{:.2f} MB\n'.format(process.memory_info().rss / 1024 / 1024))
        except KeyboardInterrupt:
            print('Skipping...')

    examples.append(wrapped)
    return wrapped

@example
def basic():
    with open(csv_path, newline='') as f:
        reader = csv.reader(f)
        for row in reader:
            pass

@example
def with_progressbar():
    num_lines = sum(1 for line in open(csv_path))

    with progressbar.ProgressBar(max_value=num_lines) as bar:
        with open(csv_path, newline='') as f:
            reader = csv.reader(f)
            for row in bar(reader):
                pass

@example
def with_headers():
    with open(csv_path, newline='') as f:
        reader = csv.DictReader(f)
        for row in reader:
            pass


#import resource
#rint('{:.2f} MB'.format(resource.getrusage(resource.RUSAGE_SELF).ru_maxrss / 1024 / 1024))

for example in examples:
    example()